checker/src/org/checkerframework/checker/index/upperbound/UpperBoundTransfer.java

```diff
@@ -124,14 +131,30 @@ private void knownToBeArrayLength(
 
         UBQualifier newInfo = UBQualifier.createUBQualifier(arrayExp, "-1");
         UBQualifier newLeft = left.glb(newInfo.plusOffset(node.getRightOperand(), atypeFactory));
+        newLeft = accountForLowerBoundAnnos(node.getRightOperand(), newLeft, arrayExp);
+
         Receiver leftRec = FlowExpressions.internalReprOf(atypeFactory, node.getLeftOperand());
         store.insertValue(leftRec, atypeFactory.convertUBQualifierToAnnotation(newLeft));
 
         UBQualifier newRight = right.glb(newInfo.plusOffset(node.getLeftOperand(), atypeFactory));
+        newRight = accountForLowerBoundAnnos(node.getLeftOperand(), newRight, arrayExp);
+
         Receiver rightRec = FlowExpressions.internalReprOf(atypeFactory, node.getRightOperand());
         store.insertValue(rightRec, atypeFactory.convertUBQualifierToAnnotation(newRight));
     }
 
+    /** If the node is NN but not POS, add an LTEL to the qual. If POS, an LTL. */
```

Comment:
Sort of. But a NonNegative could be Positive _most_ of the time, but not in every conceivable situation. Or the checker might just not be able to prove it.